### PR TITLE
Fixes #26527 - support for filtering by dashboard fields

### PR DESCRIPTION
**Pull request by iNecas**
Support for additional params in tasks index page that influence the filtering:

* result
* state
* time_horizon (e.g. H24)
* time_mode (recent/older)

Required for https://github.com/theforeman/foreman-tasks/pull/399

### CODE REVIEW

**Comment by iNecas** [comment]
All tests are :green_heart: . @adamruzicka hopefully this should not be too much of work to review: the new code should be fully covered by tests + @sharvit already confirmed it working

File: app/services/foreman_tasks/dashboard_table_filter.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      @scope = scope
+      @params = params
+    end
+
+    def scope
+      @new_scope = @scope
+      scope_by_result
+      scope_by_state
+      scope_by_time
+      @new_scope
```
[the remark this answers is not in the document]
Comment (iNecas) [reply app/services/foreman_tasks/dashboard_table_filter.rb]: I prefer keeping this explicit here (or we can rewrite it into Haskell or F# and then I will leave it out :-)

File: app/services/foreman_tasks/dashboard_table_filter.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    end
+
+    def scope_by_time
+      return if @params[:time_horizon].blank?
+      hours = @params[:time_horizon][/\A(H)(\d{1,2})$/i, 2]
+      unless hours
+        raise Foreman::Exception, 'Unexpected format of time: should be in form of "H24"'
+      end
+      timestamp = Time.now.utc - hours.to_i.hours
+      case @params[:time_mode]
```
[the remark this answers is not in the document]
Comment (iNecas) [reply app/services/foreman_tasks/dashboard_table_filter.rb]: I prefer the verbosity here, if you don't mind too much

File: app/services/foreman_tasks/dashboard_table_filter.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      @new_scope = @new_scope.where(result: @params[:result]) if @params[:result].present?
+    end
+
+    def scope_by_state
+      @new_scope = @new_scope.where(state: @params[:state]) if @params[:state].present?
+    end
+
+    def scope_by_time
+      return if @params[:time_horizon].blank?
+      hours = @params[:time_horizon][/\A(H)(\d{1,2})$/i, 2]
```
[the remark this answers is not in the document]
Comment (iNecas) [reply app/services/foreman_tasks/dashboard_table_filter.rb]: updated

File: app/services/foreman_tasks/dashboard_table_filter.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    private
+
+    def scope_by_result
+      @new_scope = @new_scope.where(result: @params[:result]) if @params[:result].present?
+    end
+
+    def scope_by_state
+      @new_scope = @new_scope.where(state: @params[:state]) if @params[:state].present?
+    end
```
[the remark this answers is not in the document]
Comment (iNecas) [reply app/services/foreman_tasks/dashboard_table_filter.rb]: good catched. DRYied it a bit

**Comment by iNecas** [comment]
Still :green_apple:

File: app/services/foreman_tasks/dashboard_table_filter.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/services/foreman_tasks/dashboard_table_filter.rb]: Good idea, I'll wait :) Just kidding, this is fine

**Comment by adamruzicka** [comment]
Thanks @iNecas & @sharvit !

**Pull request merged**